core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArrayAggregationAfterDistinct.java

```diff
@@ -0,0 +1,156 @@
+                        Symbol aggregationSymbol = symbolAggregationEntry.getKey();
+                        if (!aggregation.isDistinct()) {
+                            // recreate aggregation with distinct
+                            Symbol newArrayAggSymbol = context.getSymbolAllocator().newSymbol(aggregationSymbol);
+                            aggregationsBuilder.put(newArrayAggSymbol, new Aggregation(
+                                    aggregation.getResolvedFunction(),
+                                    aggregation.getArguments(),
+                                    true,
+                                    aggregation.getFilter(),
+                                    aggregation.getOrderingScheme(),
+                                    aggregation.getMask()));
+                            assignmentsBuilder.put(entry.getKey(), new SymbolReference(newArrayAggSymbol.getName()));
+                        }
+                        else {
+                            // already distinct so keep as is
+                            aggregationsBuilder.put(aggregationSymbol, aggregation);
+                            assignmentsBuilder.put(entry.getKey(), new SymbolReference(aggregationSymbol.getName()));
+                        }
+                        matchedAggregations.add(aggregationSymbol);
+                        matchedAssignments.add(symbol);
+                    }
+                }
+            }
+        }
+        if (!matchedAggregations.isEmpty() || !matchedAssignments.isEmpty()) {
+            // re-add not matched assignments
+            projectNode.getAssignments().entrySet().stream()
+                    .filter(entry -> !matchedAssignments.contains(entry.getKey()))
+                    .forEach(entry -> assignmentsBuilder.put(entry.getKey(), entry.getValue()));
+
+            Assignments newAssignments = assignmentsBuilder.build();
+
+            // re-add not matched or still referenced aggregations
+            aggregationNode.getAggregations().entrySet().stream()
+                    .filter(entry -> !matchedAggregations.contains(entry.getKey()) || !newAssignments.filter(matchedAggregations).isEmpty())
+                    .forEach(entry -> aggregationsBuilder.put(entry.getKey(), entry.getValue()));
+
+            AggregationNode newAggregationNode = new AggregationNode(
+                    context.getIdAllocator().getNextId(),
+                    aggregationNode.getSource(),
+                    aggregationsBuilder.buildOrThrow(),
+                    aggregationNode.getGroupingSets(),
+                    aggregationNode.getPreGroupedSymbols(),
+                    aggregationNode.getStep(),
+                    aggregationNode.getHashSymbol(),
+                    aggregationNode.getGroupIdSymbol());
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), newAggregationNode, newAssignments));
```

Comment:
I think we need to work a bit more on the approach.
1. Each time an `array_agg` aggregation is used in the `array_distinct` Projection, a new copy is created with the `distinct` property, and it is associated with a new output symbol. Instead, we should create _one_ distinct aggregation, and use it in every Projection's `array_distinct` assignment.
2. I don't think we want to end up with two versions of one aggregation: with and without the `distinct` property. If the aggregation is referenced multiple times, and not all references are `array_distinct`, then I would just bail out and leave the aggregation as it is.

Reply:
Thanks for the pointers, they are very helpful!. I've changed the approach accordingly. I've also updated the test cases (with a few TODO's left to discuss).